### ISSUE TITLE
Is there a way to access struct keys in context?

### ISSUE THREAD

**Issue by tmjoen** [markglenn/liquex#9]
Hi, and first of all thank you so much for a magnificent library!

I was wondering if it is possible to access struct keys?

I added a protocol implementation for `Strings.Char`/`to_string` to render the struct if accessed directly, but it would be nice if I could access the keys themselves as well:

```elixir
person = %Person{location: "Oslo", name: "Blabla"}
{:ok, template_ast} = Liquex.parse("Hello {{ person.location }}!")
context = Liquex.Context.new(%{"person" => person})  
{content, _context} = Liquex.render(template_ast, context)  
{["Hello ", "", "!"],
 %Liquex.Context{
   cycles: %{},
   errors: [],
   filter_module: Liquex.Filter,
   private: %{},
   render_module: nil,
   variables: %{"person" => %Person{location: "Oslo", name: "Blabla"}}
 }}
```

If I add this to `Liquex.Argument` module

```elixir
  defp do_eval(struct, [{:key, key} | tail]) when is_struct(struct) do
    atom_key = safe_to_existing_atom(key)

    struct
    |> Map.get(atom_key)
    |> do_eval(tail)
  end

  defp safe_to_existing_atom(str) do
    try do
      String.to_existing_atom(str)
    rescue
      ArgumentError -> ""
    end
  end  
```

I can make it work, but I'm not sure if this is an acceptable way of doing things? :)

If you are opposed to this, is there another way I could make this work? Extend the `do_eval` logic somehow?

Thanks again!

**Comment by markglenn** [comment]
I have gone back and forth on ideas here.  It seems inelegant to have to check for existing atoms with a rescue block, but Elixir and Erlang don't give us much choice .  My initial take was to implement a representer system that converts a struct into a map with string keys.  I don't think it's much better, but it works.

Examples can be found in the tests here: 
https://github.com/markglenn/liquex/blob/master/test/liquex/representable_test.exs

The documentation can be found here:
https://hexdocs.pm/liquex/Liquex.Represent.html#represent/2

Again, even though it can be done lazily, it is still duplication of keys.  I'm open to allowing direct access to atom keys since it's a problem I face today too.

In regards to your solution, it does work, but it misses maps that may contain atom keys.  I believe `Map.has_key?(str)` and `Map.has_key?(safe_to_existing_atom(str))` checks may be a better solution.

I will think more about it and I'm open to more suggestions.

**Comment by tmjoen** [comment]
Thanks for taking a look!

I like the `Representable` solution - only downside for me is I lose the struct info. For instance if I have a `%Global{}`: 
```elixir
defimpl String.Chars, for: Global do
  def to_string(%{type: :color, data: %{value: value}}) do
    value
  end

  def to_string(%{type: :link, data: %{type: :local, value: value}}) do
    local_url(value)
  end

  # .. etc
end
```
for calling the it directly in a template. It's a pretty nice convenience :)

**Comment by markglenn** [comment]
I've made some progress on this request.  If you look at pull request #10, it now includes a module called `Liquex.Indifferent` that does a very simple check for atom keys as a fallback if the original key (string/number/etc) isn't found.

I'm not 100% sold on it yet, but I think atom keys are such a norm in Elixir that having an implicit conversion is ok here instead of an explicit wrapper or behavior.  I'm hoping to release a new version this weekend.

I really like your solution, but I don't necessarily want to be duplicating maps and keys as a default.  Plus, Liquex lets you store anonymous functions in the context for lazy loading values.  In that case, the code to stringify keys would need to be smarter when executing those lazy functions.

**Comment by tmjoen** [comment]
Works great with your last push to the PR. Nice work!

If you'd allow me to bend your ear a final time:

I ran into a little snag with the handling of structs through the String.Chars that I mentioned in the original issue. 

I have two structs: a `%Global{}` and a `%Var{}`. Both these have a field called `type` that could be `text`/`color`/`boolean`/`markdown`/etc.

Originally I defimpl'ed them with `String.Chars:to_string`, which works fine for rendering them. The problem appeared when using them in some sort of control flow. I can't compare my `%Global{type: :boolean, data: %{value: false}}` with `false`.

If there was a protocol for `Liquex` it could work:

```elixir
defprotocol Liquex.Protocol do
  @fallback_to_any true
  @spec render(t) :: any()
  def render(value)
end

defimpl Liquex.Protocol, for: Any do
  def render(nil), do: nil
  def render(false), do: false
  def render(true), do: true
  def render(value), do: to_string(value)
end
```

This falls back to the `to_string` that you use today.

So if I change the left/right logic in `Liquex.Expression`:

```elixir
# wrapped `left` and `right` in Protocol
alias Liquex.Protocol

defp do_eval({left, op, right}), do:
    apply(Kernel, op, [Protocol.render(left), Protocol.render(right)])
```

and call the protocol for object rendering in `Liquex.Render.Object`:
```elixir
  def do_render([argument, filters: filters], %Context{} = context) do
    {result, context} =
      argument
      |> List.wrap()
      |> Argument.eval(context)
      |> Filter.apply_filters(filters, context)

    # pass result through protocol
    {Protocol.render(result), context}
  end
```

then I can have full control of my custom structs if I implement the `Liquex.Protocol` protocol:

```elixir
  defimpl Liquex.Protocol, for: Brando.Sites.Global do
    # ...
    def render(%{type: "boolean", data: data}) do
      Map.get(data, "value", false)
    end
    # ...
  end
```

Is this something you'd consider?

If not, sorry for the noise! It's just that this library has solved 99% of my problems and the only one remaining is this :smile:

**Comment by markglenn** [comment]
Can you open another issue related to that?  You can just copy/paste what you have here.  I want to keep this issue focused on the atom keys.  I can discuss more in the new issue.

**Comment by markglenn** [comment]
Just released 0.5.0 which includes this change.
